docs/aztec/protocol/trees/indexed-merkle-tree.mdx

```diff
@@ -85,85 +118,218 @@ Suppose we want to show that the value `20` doesn't exist in the tree. We just r
-  - `assert(updatedLeaf.nextVal == newLeaf.val)`
-- Replace the `updatedLeaf` in the tree: `n` hashes.
-- Hash the new leaf.
-- Add the `newLeaf` to the tree: `n` hashes.
-
-#### Number of constraints, in total:
+1. First we prove that the subtree we are inserting into consists of all empty values.
+1. We work out the root of an empty subtree, and perform an inclusion proof for an empty root, which proves that there is nothing within our subtree.
+1. We re-create our subtree within our circuit.
+1. We then use the same sibling path the get the new root of the tree after we insert the subtree.
 
-- `3n` hashes of 2 field elements (where `n` is the height of the tree).
-- `3` hashes of 3 field elements.
-- `2` range checks.
-- A handful of equality constraints.
+In the following example we insert a subtree of size 4 into our tree at step 4. above. Our subtree is greyed out as it is "pending".
+
+1. Prepare to insert subtree $[35,50,60,15]$
+
+<Image img={require("/img/indexed-merkle-tree/subtree-insert-1.png")} />
+
+2. Update low nullifier for new nullifier 35.
+
+<Image img={require("/img/indexed-merkle-tree/subtree-insert-2.png")} />
+
+3. Update low nullifier for new nullifier 50. (Notice how the low nullifier exists within our pending insertion subtree, this becomes important later).
+
+<Image img={require("/img/indexed-merkle-tree/subtree-insert-3.png")} />
+
+4. Update low nullifier for new nullifier 60.
+
+<Image img={require("/img/indexed-merkle-tree/subtree-insert-4.png")} />
+
+5. Update low nullifier for new nullifier 15.
+
+<Image img={require("/img/indexed-merkle-tree/subtree-insert-5.png")} />
+
+6. Update pointers for new nullifier 15.
+
+<Image img={require("/img/indexed-merkle-tree/subtree-insert-6.png")} />
+
+7. Insert subtree.
+
+<Image img={require("/img/indexed-merkle-tree/subtree-insert-7.png")} />
+
+#### Performance gains from subtree insertion
+
+Lets go back over the numbers:
+Insertions into a sparse nullifier tree involve 1 non membership check (254 hashes) and 1 insertion (254 hashes). If we were performing insertion for 4 values that would entail 2032 hashes.
+In the depth 32 indexed tree construction, each subtree insertion costs 1 non membership check (32 hashes), 1 pointer update (32 hashes) for each value as well as the cost of constructing and inserting a subtree (~67 hashes. Which is 327 hashes, an incredible efficiency increase.)
+
+_I am ignoring range check constraint costs as they a negligible compared to the costs of a hash_.
+
+## Performing subtree insertions in a circuit context
+
+Some fun engineering problems occur when we inserting a subtree in circuits when the low nullifier for a value exists within the subtree we are inserting. In this case we cannot perform a non membership check against the root of the tree, as our leaf that we would use for non membership has NOT yet been inserted into the tree. We need another protocol to handle such cases, we like to call these "optimistic" insertions.
```

Comment:
I still don't really like the name optimistic here 😬

Reply:
just left it as pending

Reply:
Took a me a second to catch that the name was updated to pending and not that the "fix is pending" 😆